pulsar-jms-tracing/src/main/java/com/datastax/oss/pulsar/jms/tracing/TracingUtils.java

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.jms.tracing;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import io.netty.buffer.ByteBuf;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.mledger.Entry;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.pulsar.broker.service.Consumer;
+import org.apache.pulsar.broker.service.Producer;
+import org.apache.pulsar.broker.service.ServerCnx;
+import org.apache.pulsar.broker.service.Subscription;
+import org.apache.pulsar.broker.service.Topic;
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.naming.TopicName;
+
+@Slf4j
+public class TracingUtils {
+  private static final org.slf4j.Logger traceLogger = org.slf4j.LoggerFactory.getLogger("jms-tracing");
+
+  @FunctionalInterface
+  public interface Tracer {
+    void trace(String message);
+  }
+
+  public static final Tracer SLF4J_TRACER = traceLogger::info;
+
+  private static final ObjectMapper mapper =
+      new ObjectMapper()
+          .enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
+          .enable(SerializationFeature.WRITE_NULL_MAP_VALUES);
+
+  public static final int MAX_DATA_LENGTH = 1024;
+
+  public enum TraceLevel {
+    NONE,
+    MINIMAL,
+    BASIC,
+    FULL
+  }
+
+
+  public static void trace(String message, Map<String, Object> traceDetails) {
+    trace(SLF4J_TRACER, message, traceDetails);
+  }
+
+  public static void trace(Tracer tracer, String message, Map<String, Object> traceDetails) {
+    Map<String, Object> trace = new TreeMap<>();
+    trace.put("message", message);
+    trace.put("traceDetails", traceDetails);
+
+    try {
+      String loggableJsonString = mapper.writeValueAsString(trace);
+      tracer.trace(loggableJsonString);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
```

Comment:
throwing RuntimeException in any place in the broker may lead to unpredictable behavior, like re-rending messages or leaving threads in a weird state.

I suggest to only log the problem